src/collectors/github_runners.rs

```diff
@@ -0,0 +1,180 @@
+
+#[derive(Debug, serde::Deserialize)]
+struct ApiResponse {
+    total_count: usize,
+    runners: Vec<Runner>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct Runner {
+    id: usize,
+    name: String,
+    os: String,
+    status: String,
+    busy: bool,
+    labels: Vec<Label>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct Label {
+    id: usize,
+    name: String,
+    #[serde(rename = "type")]
+    the_type: String,
+}
+
+#[derive(Clone)]
+pub struct GithubRunners {
+    //api token to use
+    token: String,
+    // repos to track gha runners
+    repos: Vec<String>,
+    // metric namespace
+    ns: String,
+    // actual metrics
+    metrics: Arc<RwLock<Vec<IntGauge>>>,
+    // default metric description
+    desc: Desc,
+}
+
+impl GithubRunners {
+    pub async fn new(config: &Config) -> Result<Self, Error> {
+        let token = config.rust_runners_token.to_string();
+        let repos: Vec<String> = config
+            .gha_runners_repos
+            .split(',')
+            .map(|v| v.trim().to_string())
+            .collect();
+
+        let ns = String::from("gha_runner");
+        let rv = Self {
+            token,
+            repos,
+            ns: ns.clone(),
+            metrics: Arc::new(RwLock::new(Vec::new())),
+            desc: Desc::new(
+                ns,
+                "GHA runner's status".to_string(),
+                Vec::new(),
+                HashMap::new(),
+            )
+            .unwrap(),
+        };
+
+        let refresh_rate = config.gha_runners_cache_refresh;
+        let mut rv2 = rv.clone();
+        tokio::spawn(async move {
+            loop {
+                if let Err(e) = rv2.update_stats().await {
+                    error!("{:#?}", e);
+                }
+
+                tokio::time::delay_for(Duration::from_secs(refresh_rate)).await;
+            }
+        });
+
+        Ok(rv)
+    }
+
+    async fn update_stats(&mut self) -> Result<(), Error> {
+        let mut gauges = Vec::new();
+        for repo in self.repos.iter() {
+            let url = String::from(GH_RUNNERS_ENDPOINT).replace("{owner_repo}", repo);
+
+            // does this token still able to query github api? (rate limit wise)
+            match is_token_flagged(&self.token).await {
+                Err(e) => {
+                    error!("checking if token is flagged: {:?}", e);
+                    continue;
+                }
+                Ok(true) => {
+                    warn!(
+                        "token: '{}' is currently flagged. skipping data gathering",
+                        &self.token
+                    );
+                    continue;
+                }
+                Ok(false) => {}
+            }
+
+            debug!("Querying gha runner's status at: {}", url);
+            let resp = http::get(&self.token, &url)
+                .send()
+                .await?
+                .json::<ApiResponse>()
+                .await?;
```

Comment:
This does not handle pagination, and it will fail to monitor some runners if we ever get more than 100 of them. [GitHub documentation on pagination](https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#pagination)

Reply:
I won't be able to set up an environment locally that will force github to send me the `Link` header. while developing this collector. (I've set up one, and found no way to fork the process to multiple instances) I'll have to follow only the docs.

having said that i was thinking in bringing in the crate https://crates.io/crates/parse_link_header into the project and let that do the parsing of the header when present. what your thought on bringing in that crate?

Reply:
Adding that dependency seems fine! Parsing that header is always a pain.

A thing you can do is setup just two runners by simply having two different installations of it, and pass `?per_page=1` to the API. That'd allow you to test pagination with just two runners.

Reply:
pagination added using `parse_link_header` crate.
also changed github api endpoint url to include `?per_page=100` by default.